processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeDeriver.java

```diff
@@ -0,0 +1,143 @@
+
+
+    public CompletionTimeDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTime processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        /**
+         * TODO HWKAMP-348
+         *
+         * Wait one retry delay before calculating trace completion time. This is not a perfect solution but
+         * there is no information in parent span whether there are any children. Solution might be calculating
+         * trace completion time for all spans (not only for root as it is now) and update duration values directly
+         * in database.
+         */
+        if (completionTimeProcessing.getNumberOfRetries() < 1) {
+            completionTimeProcessing.setNumberOfRetries(completionTimeProcessing.getNumberOfRetries() + 1);
+            return null;
+        }
+
+        Span rootSpan = completionTimeProcessing.getRootSpan();
+
+        List<Span> trace = spanCache.getTrace(tenantId, rootSpan.getTraceId());
+        Annotation lastAnnotation = extractLastAnnotation(trace);
+
+        CompletionTime completionTime = CompletionTimeUtil.spanToCompletionTime(rootSpan);
+
+        if (lastAnnotation != null) {
+            completionTime.setDuration(getTraceDuration(rootSpan, lastAnnotation.getTimestamp()));
+        }
+
+        completionTime.setProperties(extractProperties(trace));
+
+        log.debugf("SpanTraceCompletionTimeDeriver completionTime[%s]", completionTime);
+
+        return completionTime;
+    }
+
+    @Override
+    public long getRetryDelay(List<CompletionTimeProcessing> completionTimeProcessings, int retryCount) {
+        // TODO HWKAMP-348
+        return 5000;
+    }
+
+    private Long getTraceDuration(Span rootSpan, long lastAnnotationTimestamp) {
+        /**
+         * first estimate is last timestamp - first timestamp (of root span)
+         */
+        long estimatedDuration = lastAnnotationTimestamp - rootSpan.getTimestamp();
+
+        /**
+         * first estimate is only used if is bigger than root span duration. Instrumentation is
+         * responsible for setting the duration so if it is bigger number use original duration of a root span.
+         */
+        if (estimatedDuration < rootSpan.getDuration()) {
+            estimatedDuration = rootSpan.getDuration();
+        }
+
+        return estimatedDuration;
+    }
+
+    private Annotation extractLastAnnotation(List<Span> spans) {
+        Annotation lastAnnotation = null;
+
+        for (Span span: spans) {
+            for (Annotation annotation: span.getAnnotations()) {
+                if (lastAnnotation == null ||
+                        lastAnnotation.getTimestamp() < annotation.getTimestamp()) {
+                    lastAnnotation = annotation;
+                }
```

Comment:
Sorry, forgot that timestamp is recorded with the annotation -  was thinking it was on the span and therefore span timestamp + duration would mark the end time.